### ISSUE TITLE
loki sink: thread 'vector-worker' panicked at 'Batch is empty' (out-of-order event)

### ISSUE THREAD

**Issue by candlerb** [vectordotdev/vector#9279]
### Vector Version

```
vector 0.16.1 (x86_64-unknown-linux-gnu 739e878 2021-08-26)
```

Installed package `vector-0.16.1-amd64.deb` under Ubuntu 18.04

### Vector Configuration File

This is the entire config file:

```toml
# https://vector.dev/docs/reference/configuration/sources/syslog/
[sources.syslog]
  type = "syslog"
  address = "0.0.0.0:5141"
  mode = "udp"

# https://vector.dev/docs/reference/configuration/sinks/loki/
[sinks.loki]
  type = "loki"
  inputs = ["syslog"]
  endpoint = "http://localhost:3100"
  encoding.codec = "json"

  labels.job = "syslog"
  labels.source_ip = "{{ source_ip }}"
  labels.hostname = "{{ hostname }}"
  labels.facility = "{{ facility }}"
  labels.severity = "{{ severity }}"
```

### Debug Output

See https://gist.github.com/candlerb/572ec8c87055942a77e329942b8724f6

This is from recreating the problem using `systemctl edit vector.service` and

```
[Service]
Environment=RUST_BACKTRACE=full
ExecStart=
ExecStart=/usr/bin/vector -vvv
```

### Expected Behavior

Syslog message to be forwarded to loki

### Actual Behavior

Vector panics and terminates.  Without debugging, `journalctl -eu vector` shows:

```
Sep 21 09:18:20 loki vector[24493]: Sep 21 09:18:20.041  WARN sink{component_id=loki component_kind="sink" component_type=loki component_name=loki}: vector::sinks::util::buffer::loki: msg="Received out-of-order event; dropping event." internal_log_rate_secs=30
Sep 21 09:18:21 loki vector[24493]: thread 'vector-worker' panicked at 'Batch is empty', src/sinks/util/buffer/loki.rs:233:46
Sep 21 09:18:21 loki vector[24493]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Sep 21 09:18:21 loki vector[24493]: Sep 21 09:18:21.057 ERROR sink{component_id=loki component_kind="sink" component_type=loki component_name=loki}: vector::topology: An error occurred that vector couldn't handle.
Sep 21 09:18:21 loki vector[24493]: Sep 21 09:18:21.058  INFO vector: Vector has stopped.
Sep 21 09:18:21 loki vector[24493]: Sep 21 09:18:21.063  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="syslog" time_remaining="59 seconds left"
Sep 21 09:18:21 loki vector[24493]: Sep 21 09:18:21.064  INFO source{component_kind="source" component_id=syslog component_type=syslog component_name=syslog}: vector::sources::syslog: Finished sending.
Sep 21 09:18:21 loki systemd[1]: vector.service: Failed with result 'exit-code'.
```

### Example Data

A syslog message was generated using the following command: choose a timestamp slightly before the current time.

```
echo -n '<13>1 2021-09-21T09:35:00.119Z dynamicwireless.name non 2426 ID931 [exampleSDID@32473 iut="3" eventSource="Application" eventID="1011"] Try to override the THX port, maybe it will reboot the neural interface!' | nc -u localhost 5141
```

Then if necessary, repeat the command with a slightly earlier timestamp (here 1 minute earlier):

```
echo -n '<13>1 2021-09-21T09:34:00.119Z dynamicwireless.name non 2426 ID931 [exampleSDID@32473 iut="3" eventSource="Application" eventID="1011"] Try to override the THX port, maybe it will reboot the neural interface!' | nc -u localhost 5141
```

### Additional Context

Loki version is loki-2.2.0

I had previously tried messages where the timestamps were not acceptable to loki because they were too far in the past or in the future: these were rejected but did not cause vector to crash.

```
Sep 21 09:16:55 loki vector[24493]: Sep 21 09:16:55.267 ERROR sink{component_id=loki component_kind="sink" component_type=loki component_name=loki}:request{request_id=0}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 400 Bad Request"
Sep 21 09:16:55 loki vector[24493]: Sep 21 09:16:55.268 ERROR sink{component_id=loki component_kind="sink" component_type=loki component_name=loki}:request{request_id=0}: vector::sinks::util::sink: Response failed. response=Response { status: 400, version: HTTP/1.1, headers: {"content-type": "text/plain; charset=utf-8", "x-content-type-options": "nosniff", "date": "Tue, 21 Sep 2021 09:16:55 GMT", "content-length": "183"}, body: b"entry for stream '{facility=\"user\", hostname=\"dynamicwireless.name\", job=\"syslog\", severity=\"notice\", source_ip=\"127.0.0.1\"}' has timestamp too old: 2020-03-13 20:45:38.119 +0000 UTC\n" }
Sep 21 09:17:53 loki vector[24493]: Sep 21 09:17:53.499 ERROR sink{component_id=loki component_kind="sink" component_type=loki component_name=loki}:request{request_id=1}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 400 Bad Request"
Sep 21 09:17:53 loki vector[24493]: Sep 21 09:17:53.499 ERROR sink{component_id=loki component_kind="sink" component_type=loki component_name=loki}:request{request_id=1}: vector::sinks::util::sink: Response failed. response=Response { status: 400, version: HTTP/1.1, headers: {"content-type": "text/plain; charset=utf-8", "x-content-type-options": "nosniff", "date": "Tue, 21 Sep 2021 09:17:53 GMT", "content-length": "183"}, body: b"entry for stream '{facility=\"user\", hostname=\"dynamicwireless.name\", job=\"syslog\", severity=\"notice\", source_ip=\"127.0.0.1\"}' has timestamp too new: 2021-09-21 10:17:00.119 +0000 UTC\n" }
```

If I tcpdump localhost port 3100, I can see the successful message being delivered, but when the out-of-order timestamp message is submitted to vector, there is no activity to loki.

It appears that this out-of-order event is being dropped by vector itself before forwarding to loki, and generating the log message itself [here](https://github.com/vectordotdev/vector/blob/v0.16.1/src/sinks/util/buffer/loki.rs#L157).

(This has moved to [src/internal_events/loki.rs](https://github.com/vectordotdev/vector/blob/master/src/internal_events/loki.rs#L40) in the master branch)

### References

n/a

**Comment by candlerb** [comment]
Sorry, I think this is a duplicate of #9124.  However, here I give an easy way to reproduce without having to generate any substantial level of load.  Just generate two events with the same labels and T2 < T1.

Also worth mentioning #9223, although that's a different issue (*when* Loki accepts out-of-order writes, *then* Vector won't need to enforce this check)
